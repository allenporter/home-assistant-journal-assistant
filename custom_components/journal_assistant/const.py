"""Constants for journal_assistant."""

DOMAIN = "journal_assistant"

VISION_MODEL_NAME = "gemini-1.5-flash"
CONF_API_KEY = "api_key"
CONF_MEDIA_SOURCE = "media_source"
CONF_NOTES = "notes"
DEFAULT_NOTES = [
    "Daily",
    "Weekly",
    "Monthly",
]
DEFAULT_NOTE_NAME = "Journal"

CONF_CONFIG_ENTRY_ID = "config_entry_id"
