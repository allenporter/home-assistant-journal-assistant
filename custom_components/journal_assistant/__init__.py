"""journal_assistant custom component."""

from __future__ import annotations

import logging

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant

from .const import DOMAIN
from .services import async_register_services
from .llm import async_register_llm_apis
from .types import JournalAssistantConfigEntry
from .storage import create_vector_db

__all__ = [
    "DOMAIN",
]

_LOGGER = logging.getLogger(__name__)


PLATFORMS = (Platform.CALENDAR, Platform.SENSOR)


async def async_setup_entry(
    hass: HomeAssistant, entry: JournalAssistantConfigEntry
) -> bool:
    """Set up a config entry."""
    entry.runtime_data = await create_vector_db(hass, entry)
    await hass.config_entries.async_forward_entry_setups(
        entry,
        platforms=PLATFORMS,
    )
    async_register_services(hass)
    await async_register_llm_apis(hass, entry)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(
        entry,
        PLATFORMS,
    )
