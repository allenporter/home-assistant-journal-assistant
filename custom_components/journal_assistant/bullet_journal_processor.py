"""Processes a media source as a Bullet Journal.

This module is used to process a media source as a Bullet Journal. It extracts
text from the media source using a vision model and updates the vector database
with new journal entries.
"""

from typing import Any
import logging

from homeassistant.core import HomeAssistant, CALLBACK_TYPE, Event

from .media_source_listener import async_create_media_source_listener
from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


DATA_KEY = "bullet_journal_processor"


def async_register(
    hass: HomeAssistant, config_entry_id: str, media_source_domain: str
) -> CALLBACK_TYPE:
    """Register the media source listener."""
    listener_id = f"{DOMAIN}_{config_entry_id}"

    processor = BulletJournalProcessor(hass)
    cancel = hass.bus.async_listen(listener_id, processor.handle_event)
    unsub = async_create_media_source_listener(hass, media_source_domain, listener_id)

    def cleanup() -> None:
        """Clean up the media source listener."""
        cancel()
        unsub()

    return cleanup


class BulletJournalProcessor:
    """Library for processing a media source as a Bullet Journal."""

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the Bullet Journal Processor."""
        self._hass = hass

    async def handle_event(self, event: Event[dict[str, Any]]) -> None:
        """Handle an updated media source object."""
        _LOGGER.info("Processing media source event: %s", event)
