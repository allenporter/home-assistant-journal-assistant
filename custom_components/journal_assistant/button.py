"""Journal assistant button platform."""

import datetime
import logging


from homeassistant.const import EntityCategory
from homeassistant.core import HomeAssistant
from homeassistant.components.button import ButtonEntity
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import DOMAIN
from .types import JournalAssistantConfigEntry

_LOGGER = logging.getLogger(__name__)

UPDATE_INTERVAL = datetime.timedelta(minutes=15)


async def async_setup_entry(
    hass: HomeAssistant,
    entry: JournalAssistantConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the journal sensor component."""
    async_add_entities([ProcessMediaButtonEntity(entry)], True)


class ProcessMediaButtonEntity(ButtonEntity):
    """A button that allows manual processing of media."""

    _attr_has_entity_name = True
    _attr_entity_category = EntityCategory.DIAGNOSTIC

    def __init__(self, entry: JournalAssistantConfigEntry) -> None:
        """Initialize the process media button entity."""
        self._attr_unique_id = f"{entry.entry_id}-process-media"
        self._processor = entry.runtime_data.media_source_processor
        self._attr_name = "Process Media"
        self._attr_device_info = {
            "identifiers": {(DOMAIN, entry.entry_id)},
            "name": entry.title,
        }

    async def async_press(self) -> None:
        """Handle the button press."""
        self.hass.async_create_task(
            self._processor.async_process_media(datetime.datetime.now())
        )
