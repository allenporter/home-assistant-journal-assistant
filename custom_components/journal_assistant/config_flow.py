"""Config flow for journal_assistant integration."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, cast

import voluptuous as vol

from homeassistant.const import CONF_NAME
from homeassistant.helpers import (
    config_validation as cv,
    selector,
)
from homeassistant.helpers.schema_config_entry_flow import (
    SchemaConfigFlowHandler,
    SchemaFlowFormStep,
)

from .const import (
    DOMAIN,
    CONF_NOTES,
    DEFAULT_NOTES,
    CONF_API_KEY,
)


CONFIG_FLOW = {
    "user": SchemaFlowFormStep(
        vol.Schema(
            {
                vol.Required(CONF_NAME): cv.string,
                vol.Required(CONF_API_KEY): cv.string,
                vol.Required(
                    CONF_NOTES, default="\n".join(DEFAULT_NOTES)
                ): selector.TextSelector(selector.TextSelectorConfig(multiline=True)),
            }
        )
    )
}

OPTIONS_FLOW = {
    "init": SchemaFlowFormStep(),
}


class JournalAssistantConfigFlowHandler(SchemaConfigFlowHandler, domain=DOMAIN):
    """Handle a config flow for Switch as X."""

    config_flow = CONFIG_FLOW
    options_flow = OPTIONS_FLOW

    VERSION = 1
    MINOR_VERSION = 1

    def async_config_entry_title(self, options: Mapping[str, Any]) -> str:
        """Return config entry title."""
        return cast(str, options[CONF_NAME])
