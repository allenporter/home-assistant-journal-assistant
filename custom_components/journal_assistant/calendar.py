"""A journal calendar component."""

import datetime
import logging
import slugify

from ical.calendar import Calendar
from ical.journal import Journal
from ical.timeline import generic_timeline

from homeassistant.components.calendar import CalendarEntity, CalendarEvent
from homeassistant.const import CONF_NAME
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.util import dt as dt_util

from .storage import load_journal_entries

_LOGGER = logging.getLogger(__name__)

UPDATE_INTERVAL = datetime.timedelta(minutes=15)


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the journal calendar component."""
    _LOGGER.debug("Setting up journal calendar component")
    entries = await load_journal_entries(hass, entry)
    for journal_name, calendar in entries.items():
        async_add_entities([JournalCalendar(entry, journal_name, calendar)])


class JournalCalendar(CalendarEntity):
    """A journal calendar component."""

    _attr_has_entity_name = True

    def __init__(
        self, entry: ConfigEntry, journal_name: str, calendar: Calendar
    ) -> None:
        """Initialize the journal calendar component."""
        self._attr_unique_id = f"{entry.entry_id}-{slugify.slugify(journal_name)}"
        self._entry = entry
        self._attr_name = entry.options[CONF_NAME] + ": " + journal_name
        self._calendar = calendar
        self._event: CalendarEvent | None = None

    @property
    def event(self) -> CalendarEvent | None:
        """Return the events of the calendar."""
        return self._event

    async def async_get_events(
        self,
        hass: HomeAssistant,
        start_date: datetime.datetime,
        end_date: datetime.datetime,
    ) -> list[CalendarEvent]:
        """Return calendar events within a datetime range."""
        events = generic_timeline(
            self._calendar.journal, start_date.tzinfo or dt_util.UTC
        ).overlapping(
            start_date,
            end_date,
        )
        return [_get_calendar_event(event) for event in events]

    async def async_update(self) -> None:
        """Update entity state with the next upcoming event."""
        now = dt_util.now()
        events = generic_timeline(
            self._calendar.journal, now.tzinfo or dt_util.UTC
        ).active_after(now)
        if event := next(events, None):
            self._event = _get_calendar_event(event)
        else:
            self._event = None


def _get_calendar_event(event: Journal) -> CalendarEvent:
    """Return a CalendarEvent from an API event."""
    start: datetime.datetime | datetime.date
    end: datetime.datetime | datetime.date
    if isinstance(event.start, datetime.datetime):
        start = dt_util.as_local(event.start)
        end = start + event.computed_duration
    else:
        start = event.start
        end = start + event.computed_duration

    return CalendarEvent(
        summary=event.summary or "",
        start=start,
        end=end,
        description=event.description,
        uid=event.uid,
    )
