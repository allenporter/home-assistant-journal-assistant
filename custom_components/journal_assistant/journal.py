"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging

from ical.calendar import Calendar
from ical.journal import Journal
from .model import JournalPage


_LOGGER = logging.getLogger(__name__)


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        dated_content[note_date].append(f"- {note.content}")
    return dated_content


def journal_from_yaml(storage_dir: Path) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    journal_names = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    _LOGGER.debug("Journal names: %s", journal_names)

    journals = {}

    keys = list(journal_names)
    keys.sort()
    for journal_name in keys:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, journal_name)

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            dated_content.update(get_dated_content(page))

        if journal_name not in journals:
            journals[journal_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[journal_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.summary = f"{journal_name} {date}"
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[journal_name] = calendar

    return journals
