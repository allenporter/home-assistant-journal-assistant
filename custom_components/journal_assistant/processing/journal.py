"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging
import hashlib
from typing import cast

from ical.calendar import Calendar
from ical.journal import Journal

from .model import JournalPage

_LOGGER = logging.getLogger(__name__)

__all__ = [
    "journal_from_yaml",
]


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        content = ""
        if note.content:
            if note.status:
                content = f"- ({note.status}) {note.content}"
            else:
                content = f"- {note.content}"
        if note.entries:
            if content:
                content += "\n"
            content += "\n".join(f"  - {entry}" for entry in note.entries if entry)
        dated_content[note_date].append(content)
    return dated_content


def write_content(content: str, filename: Path) -> None:
    """Write content to a file."""
    with filename.open("w") as file:
        file.write(content)


def write_journal_page_yaml(
    storage_dir: Path,
    note_name: str,
    page: JournalPage,
) -> None:
    """Write a journal page yaml string to a yaml file."""
    if "-" not in note_name:
        raise ValueError(f"Note name must contain a '-' character: {note_name}")
    if not page.filename.startswith(note_name):
        raise ValueError(
            f"Note name must match page filename: {note_name} != {page.filename}"
        )
    storage_dir.mkdir(parents=True, exist_ok=True)
    filename = storage_dir / f"{note_name}.yaml"
    _LOGGER.debug("Writing journal page to %s", filename)
    content = cast(str, page.to_yaml())
    write_content(content, filename)


def journal_from_yaml(
    storage_dir: Path,
    allowed_notes: set[str],
    default_note_name: str,
) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    filename_prefixes = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    note_names = list(filename_prefixes)
    note_names.sort()
    _LOGGER.debug("Journal names: %s", note_names)

    journals = {}
    for note_name in note_names:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, note_name)

        # Allow notes to have their own calendar entry if in the list of allowed notes
        key_name = note_name if note_name in allowed_notes else default_note_name

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            for date, content_list in get_dated_content(page).items():
                if date not in dated_content:
                    dated_content[date] = []
                dated_content[date].extend(content_list)

        if key_name not in journals:
            journals[key_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[key_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.uid = hashlib.sha256(f"{note_name}-{date}".encode()).hexdigest()
            journal.summary = f"{note_name} {date}"
            journal.categories = [key_name]
            if note_name not in allowed_notes:
                journal.categories.append(note_name)
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[key_name] = calendar

    return journals
