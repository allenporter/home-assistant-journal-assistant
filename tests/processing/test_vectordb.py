"""Test loading the vector DB."""

from pathlib import Path
from collections.abc import Generator
from typing import Any
from unittest.mock import patch
import hashlib
import tempfile
import datetime

import chromadb
import numpy as np
from syrupy import SnapshotAssertion
import pytest

from custom_components.journal_assistant.const import DOMAIN
from custom_components.journal_assistant.processing import vectordb
from custom_components.journal_assistant.processing.journal import journal_from_yaml


@pytest.fixture(autouse=True)
def mock_disable_posthog() -> Generator[None, None, None]:
    """Disable chromadb posthog since it starts daemon threads during tests."""
    with patch(
        "chromadb.telemetry.product.posthog.Posthog",
        autospec=True,
    ):
        yield


@pytest.fixture(name="storage_path")
def mock_vectordb_storage_path() -> Generator[Path, None, None]:
    """Fake out the storage path to load from the fixtures directory."""
    with tempfile.TemporaryDirectory() as tmpdir:
        yield Path(tmpdir)


class FakeEmbeddingFunction(chromadb.EmbeddingFunction):
    """Fake embedding function for testing."""

    embeds: int = 0

    def __call__(self, input: Any) -> chromadb.Embeddings:
        result: chromadb.Embeddings = []
        for item in input:
            self.embeds += 1
            result.append(
                np.array(
                    [ord(c) for c in hashlib.sha256(item.encode()).hexdigest()][0:3]
                )
            )
        return result


@pytest.fixture(name="embedding_function")
def mock_embedding_function() -> Generator[FakeEmbeddingFunction, None, None]:
    """Fixture to mock the embedding function."""
    fake_embedding = FakeEmbeddingFunction()
    with patch(
        f"custom_components.{DOMAIN}.processing.vectordb.google_embedding_function.GoogleGenerativeAiEmbeddingFunction",
        return_value=fake_embedding,
    ):
        yield fake_embedding


def test_vectordb_loading(
    storage_path: Path,
    embedding_function: FakeEmbeddingFunction,
    snapshot: SnapshotAssertion,
) -> None:
    """Test parsing a journal page."""

    entries = journal_from_yaml(Path("tests/fixtures"), {"Daily", "Monthly"}, "Journal")
    assert len(entries) == 3
    assert entries.keys() == {"Daily", "Journal", "Monthly"}

    # Add the first entry to the index
    first_calendar = next(iter(entries.values()))

    db = vectordb.VectorDB(storage_path, "12345")
    db.upsert_index(
        [vectordb.create_indexable_document(entry) for entry in first_calendar.journal]
    )
    assert embedding_function.embeds == 4

    # Add the rest, which skips the duplicate
    db.upsert_index(
        [
            vectordb.create_indexable_document(entry)
            for calendar in entries.values()
            for entry in calendar.journal
        ]
    )
    assert embedding_function.embeds == 7

    assert db.count() == 7

    results = db.query(vectordb.QueryParams(query="tree", num_results=5))
    assert len(results) == 5
    assert results[0] == snapshot

    # Limit results to a single date
    results = db.query(
        vectordb.QueryParams(
            query="gifts",
            category="Daily",
            date_range=(datetime.date(2023, 12, 21), datetime.date(2023, 12, 21)),
            num_results=5,
        )
    )
    assert len(results) == 1
    assert results[0] == snapshot
