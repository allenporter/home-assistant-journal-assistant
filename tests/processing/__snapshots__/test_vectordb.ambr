# serializer version: 1
# name: test_vectordb_loading
  dict({
    'category': 'Daily',
    'content': '''
      categories:
      - Daily
      description: '- clean garage'
      dtstart: 2023-12-22
      summary: Daily 2023-12-22
  
    ''',
    'date': 1703232000.0,
    'id': '3ac14d0bb8c28ac12733b156a5b96d6af62d3c3be63d488aa763a99637a407ce',
    'name': 'Daily 2023-12-22',
    'score': 5.000000000000001,
  })
# ---
# name: test_vectordb_loading.1
  dict({
    'category': 'Daily',
    'content': '''
      categories:
      - Daily
      description: '- cardboard breakdown
      
        - (migrated) Bowling w/ Q
      
        - (completed) flux-local helm
      
        - todo urls?
      
        - (migrated) gifts plan
      
        - windows xmas lights
      
        - (migrated) fitbit python'
      dtstart: 2023-12-21
      summary: Daily 2023-12-21
  
    ''',
    'date': 1703145600.0,
    'id': 'caba4b5990a778e89764bdb09f6902d6fc68a48d8fe96c1d9bbe4424d00af930',
    'name': 'Daily 2023-12-21',
    'score': 2541.0,
  })
# ---
